File: src/services/agent.ts
import { invoke } from '@tauri-apps/api/core'

import type {
  AgentConfigInput,
  AgentInfo,
  AgentToolPermission,
  AgentToolPermissionInput,
  AgentType,
  ToolGroup,
  ToolInfo,
  ToolPermission,
} from '../types/chat'

export function listAgents(): Promise<AgentInfo[]> {
  return invoke<AgentInfo[]>('agent_list')
}

export function getAgent(agentId: AgentType): Promise<AgentInfo> {
  return invoke<AgentInfo>('agent_get', { agentId })
}

export function createAgent(input: AgentConfigInput): Promise<AgentInfo> {
  return invoke<AgentInfo>('agent_create', { input })
}

export function updateAgent(agentId: AgentType, input: AgentConfigInput): Promise<AgentInfo> {
  return invoke<AgentInfo>('agent_update', { agentId, input })
}

export function deleteAgent(agentId: AgentType): Promise<void> {
  return invoke<void>('agent_delete', { agentId })
}

/** 工具目录，按工具组、id 升序。 */
export function listTools(): Promise<ToolInfo[]> {
  return invoke<ToolInfo[]>('tool_list')
}

/** 全部工具组及组内工具数量。 */
export function listToolGroups(): Promise<ToolGroup[]> {
  return invoke<ToolGroup[]>('get_all_tool_groups')
}

/** 工具目录全集 + 该 Agent 的生效级别（未配置为 deny）。 */
export function getAgentToolPermissions(agentId: AgentType): Promise<AgentToolPermission[]> {
  return invoke<AgentToolPermission[]>('agent_get_tool_permissions', { agentId })
}

export function setAgentToolPermission(
  agentId: AgentType,
  toolId: string,
  permission: ToolPermission,
): Promise<AgentToolPermission> {
  return invoke<AgentToolPermission>('agent_set_tool_permission', { agentId, toolId, permission })
}

/** 覆盖式批量设置工具权限。 */
export function setAgentToolPermissions(
  agentId: AgentType,
  items: AgentToolPermissionInput[],
): Promise<AgentToolPermission[]> {
  return invoke<AgentToolPermission[]>('agent_set_tool_permissions', { agentId, items })
}

/** 纯查询某 Agent 对某工具的生效级别。 */
export function resolveAgentToolPermission(
  agentId: AgentType,
  toolId: string,
): Promise<ToolPermission> {
  return invoke<ToolPermission>('agent_resolve_tool_permission', { agentId, toolId })
}

File: src-tauri/src/controller/agent.rs
//! Agent 与工具命令，仅做参数透传。

use crate::interface::agent::{
    AgentConfigInput, AgentInfo, AgentToolPermission, AgentToolPermissionInput, ToolGroup, ToolInfo,
    ToolPermission,
};
use crate::interface::error::ApiError;
use crate::AppState;

#[tauri::command]
pub fn agent_list(state: tauri::State<'_, AppState>) -> Result<Vec<AgentInfo>, ApiError> {
    state.agent.list_agents()
}

#[tauri::command]
pub fn agent_get(state: tauri::State<'_, AppState>, agent_id: i64) -> Result<AgentInfo, ApiError> {
    state.agent.get_agent(agent_id)
}

#[tauri::command]
pub fn agent_create(
    state: tauri::State<'_, AppState>,
    input: AgentConfigInput,
) -> Result<AgentInfo, ApiError> {
    state.agent.create_agent(input)
}

#[tauri::command]
pub fn agent_update(
    state: tauri::State<'_, AppState>,
    agent_id: i64,
    input: AgentConfigInput,
) -> Result<AgentInfo, ApiError> {
    state.agent.update_agent(agent_id, input)
}

#[tauri::command]
pub fn agent_delete(state: tauri::State<'_, AppState>, agent_id: i64) -> Result<(), ApiError> {
    state.agent.delete_agent(agent_id)
}

#[tauri::command]
pub fn tool_list(state: tauri::State<'_, AppState>) -> Result<Vec<ToolInfo>, ApiError> {
    state.agent.list_tools()
}

#[tauri::command]
pub fn get_all_tool_groups(state: tauri::State<'_, AppState>) -> Result<Vec<ToolGroup>, ApiError> {
    state.agent.list_tool_groups()
}

#[tauri::command]
pub fn agent_get_tool_permissions(
    state: tauri::State<'_, AppState>,
    agent_id: i64,
) -> Result<Vec<AgentToolPermission>, ApiError> {
    state.agent.get_tool_permissions(agent_id)
}

#[tauri::command]
pub fn agent_set_tool_permission(
    state: tauri::State<'_, AppState>,
    agent_id: i64,
    tool_id: String,
    permission: ToolPermission,
) -> Result<AgentToolPermission, ApiError> {
    state.agent.set_tool_permission(agent_id, &tool_id, permission)
}

#[tauri::command]
pub fn agent_set_tool_permissions(
    state: tauri::State<'_, AppState>,
    agent_id: i64,
    items: Vec<AgentToolPermissionInput>,
) -> Result<Vec<AgentToolPermission>, ApiError> {
    state.agent.set_tool_permissions(agent_id, items)
}

#[tauri::command]
pub fn agent_resolve_tool_permission(
    state: tauri::State<'_, AppState>,
    agent_id: i64,
    tool_id: String,
) -> Result<ToolPermission, ApiError> {
    state.agent.resolve_tool_permission(agent_id, &tool_id)
}
